### ISSUE TITLE
Playing "The King" engine and personalities with ChessBridge

### ISSUE THREAD

**Issue by gotofan** [Xyrus2000/ChessBridge#5]
I've followed instructions and have an alternative engine working as the "core" of Chessmaster GME (11th Edition). Now that the original King engine file has been renamed and the ChessBridge executable put in its place as TheKing350.exe, I can still start a new game against any one of the default Chessmaster personalities. Since other engines don't recognize Chessmaster personalities, my assumption is that it's playing something other than Chessmaster's The King. I'd like to continue using another engine with ChessBridge, but also play the real King engine with all its personalities whenever I like. How can this be accomplished?

**Comment by gotofan** [comment]
I've a discovered possible valid scenario, using Stockfish 14 32bit, Polyglot & ChessBridge to not only get Stockfish running in Chessmaster XI as a playing engine, but also as a reliable analysis engine:

I placed the Stockfish engine (Stockfish_14_32bit.exe, in my case), Polyglot and ChessBridge executables, and their respective config files in the same directory (outside the Chessmaster installation tree). Also, there are several polyglot.ini files tailored to Stockfish on the Internet. The chessbridge.cfg file simply requires the three entries, as described in the documentation:
mode=
gui=chessmaster
engine=polyglot.exe

Then fire up the Chessmaster, follow the standard procedure for importing a Winboard engine but enter only the full executable path of ChessBridge.exe. I named the new engine Stockfish_14. Properly configured chessbrige and polyglot config files will take care of the rest. Exit Chessmaster. (Hint: My verson of Ployglot might be confused by forward slashes in pathnames appearing in polyglot.ini, since it's being run on Windows. Maybe others are like this?)
Stockfish is now available as an opponent for you, or another engine in Chessmaster.

If you want to make Stockfish as the default engine for Chessmaster's analysis:
You will need to execute a text editor, like Notepad, as Administrator (right-click, Run As Administrator). Then open your Chessmaster account's .ini file (e.g. "C:\Program Files (x86)\Ubisoft\Chessmaster Grandmaster Edition\Data\Users\Playa\Playa.ini"). In that .ini file, search for a section [Mentor]. Add a line, making sure it is squarely within the [Mentor] section:
personality=Stockfish_14

Stockfish_14 stands for whatever exact name was given, back when importing ChessBridge.exe into Chessmaster. It can, of course, be named something else suited to taste.
Save changes to the file. Next time you analyze a game in Chessmaster, Stockfish should be the one doing the analysis. Can be confirmed by checking to see that the ChessBridge, Polyglot and Stockfish executables got loaded in Windows Task Manager.

That's it. Thanks to EscherehcsE on the chess.com forums for pointing out the required changes to the [Mentor] section in the .ini file. My limited testing has shown this configuration to be stable.

I believe this method may confer some small advantage over replacing The King executable: TheKing*.exe remain in-situ, while Stockfish is automatically summoned for any game analysis. It is then unloaded, then back to a pristine Chessmaster environment.

A few more brief notes:
The 32-bits version of Stockfish was used since, in one analysis test, game.exe (also found in the Chessmaster installation directory) hangs when compiling natural-language text, after Stockfish x64 is unloaded. At least on my setup it is consistent.
Stockfish 15 (binaries only for 64-bits) causes a crash in game.exe at exactly the same juncture, even when all executables belonging to the Chessmaster program were tried in Vista compatibility mode.
This may work with other releases of Chessmaster, but was only tested on the Grandmaster (a.k.a. 11th) Edition.
This may work with other UCI and—or other Winboard engines, but I have not tested any other as yet.
All "testing" was done on an AMD A8-based machine with Windows 7 SP1 x86_64, running Chessmaster XI updated to version 1.02.

Maybe the mystery of 64-bits will one day be solved...Game analysis works fine with the 64-bits version of The King, but Stockfish x64 falters regularly when analyzing the same game. Likely a BEX (buffer overflow exception) in these cases.

And thanks Xyrus2000, for ChessBridge.

Added: I can confirm the problem with (at least) Stockfish 15 x64 is a buffer overflow exception (BEX). Stockfish 14 does not give this indication; it only hangs on my machine.

**Comment by Xyrus2000** [comment]
I thought I had sent an email answering your question but I
apparently never hit send. :P

You are correct. Other engines do not recognize Chessmaster
personalities as those configuration options are specific to The King. In
addition, there is the OPK issue so bringing the king as an external engine
is a bit of pain without hacking the engine (though other than that, you'd
just use the personality dumper and feed the arguments into the engine). It
would take some work but you could work around these issues with a pass
through program similar to ChessBridge just making a bridge between
the Chessmaster Winboard interface and The King..

**Comment by Xyrus2000** [comment]
In that scenario, that should work.

I haven't dived into the mentor system to see how it invokes the analysis,
but if it is trying to share memory space with a 64-bit application then
it's not going to work. Unless ubisoft releases a 64 bit version of
Chessmaster then the mentor will likely not work with a 64-bit executable.
:P

**Comment by gotofan** [comment]
Thanks for the reply. Your theory is very likely correct: I've updated my above post to indicate that, after analysis with Stockfish 15 x64 at least, it's a buffer overflow exception (a window actually comes up with BEX error message). Chessmaster, using Stockfish 14 x64, hangs at the exact same time on the same .CMG file, but without spewing a helpful error message (it just hangs). So it's probably safe to assume that the data structures used by TheKing350_64 were coded to strictly accommodate the 32-bitness of its native Chessmaster interface.
As a point of minor interest: I set up a six-game chess engine tournament in Shredder, between Stockfish 14 x64 and Stockfish 14 32-bit. Both engines were geared to precisely the same settings, and each side was given 30 minutes. The latter won the series 3.5 to 2.5.
